crates/sim2h/src/lib.rs

```diff
@@ -578,40 +587,75 @@ impl Sim2h {
-                        request_id: for_agent_id.clone().into(),
-                        space_address: space_address.clone(),
-                        provider_agent_id: random_agent.clone(),
-                        entry_address: entry_address.clone(),
-                        aspect_address_list: Some(aspect_address_list.clone()),
-                    }));
-                debug!("SENDING fetch with request ID: {:?}", wire_message);
-                self.send(random_agent.clone(), random_url.clone(), &wire_message);
+                if let Some(random_agent) = self.get_random_agent_not_missing_aspects(
+                    entry_address,
+                    aspect_address_list,
+                    &agent_pool,
+                    &space_address,
+                ) {
+                    debug!(
+                        "FETCHING missing contents from RANDOM AGENT: {}",
+                        random_agent
+                    );
+
+                    let maybe_url = self.lookup_joined(&space_address, &random_agent);
+                    if maybe_url.is_none() {
+                        error!("Could not find URL for randomly selected agent. This should not happen!");
+                        return;
+                    }
+                    let random_url = maybe_url.unwrap();
+
+                    let wire_message = WireMessage::Lib3hToClient(Lib3hToClient::HandleFetchEntry(
+                        FetchEntryData {
+                            request_id: for_agent_id.clone().into(),
+                            space_address: space_address.clone(),
+                            provider_agent_id: random_agent.clone(),
+                            entry_address: entry_address.clone(),
+                            aspect_address_list: Some(aspect_address_list.clone()),
+                        },
+                    ));
+                    debug!("SENDING fetch with request ID: {:?}", wire_message);
+                    self.send(random_agent.clone(), random_url.clone(), &wire_message);
+                } else {
+                    warn!("Could not find an agent that has any of the missing aspects. Trying again later...")
+                }
             }
         }
     }
 
+    fn get_random_agent_not_missing_aspects(
+        &self,
+        entry_hash: &EntryHash,
+        aspects: &Vec<AspectHash>,
+        agent_pool: &Vec<AgentId>,
+        space_address: &SpaceHash,
+    ) -> Option<AgentId> {
+        let agents_not_missing_entry: Vec<&AgentId> = {
+            let space_lock = self.spaces.get(space_address)?.read();
+            agent_pool
+                .iter()
+                // We ignore all agents that are missing all of the same aspects as well since
+                // they can't help us.
+                .filter(|a| !space_lock.agent_is_missing_all_aspects(*a, entry_hash, aspects))
+                .collect()
+        };
```

Comment:
This could get slow as we scale up the number of agents, since we are iterating over all agents and calculating which are missing aspects. Since we ultimately only want one agent that meets the criteria, a faster way would be to generate a random permutation of vec indices so we can iterate over the agents randomly, and return the first one that doesn't have missing aspects.

Can use: https://docs.rs/rand/0.6.5/rand/seq/trait.SliceRandom.html#tymethod.shuffle